File: amazon-scraper/python/reviews.py
import urllib.parse
import csv
import re
import requests
from bs4 import BeautifulSoup

# Configuration
token = "<SDO-token>"
asin = "B07TCJS1NS"  # Change this to any product ASIN
geocode = "us"

# Build product URL from ASIN
target_url = f"https://www.amazon.com/dp/{asin}"

# Make API request
targetUrl = urllib.parse.quote(target_url)
apiUrl = "https://api.scrape.do/?token={}&url={}&geoCode={}".format(token, targetUrl, geocode)
response = requests.request("GET", apiUrl)

soup = BeautifulSoup(response.text, "html.parser")

# Extract review data
reviews = []
for review in soup.find_all("li", {"data-hook": "review"}):
    # Get star rating
    rating_elem = review.find("i", {"data-hook": "review-star-rating"}) or review.find("i", class_=re.compile(r"a-icon-star"))
    rating = rating_elem.find("span", class_="a-icon-alt").text.split()[0] if rating_elem else "N/A"

    # Get review date (remove country prefix)
    date_elem = review.find("span", {"data-hook": "review-date"})
    date = re.sub(r"Reviewed in .* on ", "", date_elem.text) if date_elem else "N/A"

    # Get review content
    content_elem = review.find("span", {"data-hook": "review-body"})
    content = content_elem.get_text(strip=True) if content_elem else "N/A"

    # Get helpful votes count
    helpful_elem = review.find("span", {"data-hook": "helpful-vote-statement"})
    helpful = re.findall(r'\d+', helpful_elem.text)[0] if helpful_elem and re.findall(r'\d+', helpful_elem.text) else "0"

    reviews.append({
        "review_id": review.get("id", ""),
        "rating": rating,
        "date": date,
        "content": content,
        "helpful": helpful
    })

# Export to CSV
csv_file = "reviews.csv"
headers = ["review_id", "rating", "date", "content", "helpful"]

with open(csv_file, "w", newline="", encoding="utf-8") as f:
    writer = csv.DictWriter(f, fieldnames=headers)
    writer.writeheader()
    writer.writerows(reviews)

print(f"Found {len(reviews)} reviews")
print(f"Data exported to {csv_file}")

File: amazon-scraper/node.js/reviews.js
const axios = require('axios');
const cheerio = require('cheerio');
const createCsvWriter = require('csv-writer').createObjectCsvWriter;

// Configuration
const token = "<SDO-token>";
const asin = "B07TCJS1NS"; // Change this to any product ASIN
const geocode = "us";

// Build product URL from ASIN
const targetUrl = `https://www.amazon.com/dp/${asin}`;

// Make API request
const encodedUrl = encodeURIComponent(targetUrl);
const apiUrl = `https://api.scrape.do/?token=${token}&url=${encodedUrl}&geoCode=${geocode}`;

async function scrapeReviews() {
    const response = await axios.get(apiUrl);
    const $ = cheerio.load(response.data);

    const reviews = [];

    $('li[data-hook="review"]').each((index, element) => {
        const review = $(element);

        // Get star rating
        const ratingElem = review.find('i[data-hook="review-star-rating"]').first() ||
                          review.find('i[class*="a-icon-star"]').first();
        const ratingText = ratingElem.find('span.a-icon-alt').text();
        const rating = ratingText ? ratingText.split(' ')[0] : 'N/A';

        // Get review date (remove country prefix)
        const dateElem = review.find('span[data-hook="review-date"]');
        const dateText = dateElem.text();
        const date = dateText ? dateText.replace(/Reviewed in .* on /, '') : 'N/A';

        // Get review content
        const contentElem = review.find('span[data-hook="review-body"]');
        const content = contentElem.text().trim() || 'N/A';

        // Get helpful votes count
        const helpfulElem = review.find('span[data-hook="helpful-vote-statement"]');
        const helpfulText = helpfulElem.text();
        const helpfulMatch = helpfulText.match(/\d+/);
        const helpful = helpfulMatch ? helpfulMatch[0] : '0';

        reviews.push({
            review_id: review.attr('id') || '',
            rating: rating,
            date: date,
            content: content,
            helpful: helpful
        });
    });

    // Export to CSV
    const csvWriter = createCsvWriter({
        path: 'reviews.csv',
        header: [
            { id: 'review_id', title: 'review_id' },
            { id: 'rating', title: 'rating' },
            { id: 'date', title: 'date' },
            { id: 'content', title: 'content' },
            { id: 'helpful', title: 'helpful' }
        ]
    });

    await csvWriter.writeRecords(reviews);

    console.log(`Found ${reviews.length} reviews`);
    console.log('Data exported to reviews.csv');
}

scrapeReviews().catch(console.error);
